### PR TITLE
Tensor Parallelism

### PR DESCRIPTION
**Pull request by eitanturok**
Implement Tensor Parallelism (TP) in foundry.

To do:
- [x] Make tp_strategy registry
- [x] Check TP: if world_size == 1 or training an MoE, don't apply TP
- [x] Get the same loss with fsdp and fsdp-tp 
- [x] Add Tests

**Updates:**
I compared training 125m param models for 100 steps on c4 with tp-fsdp VS fsdp.
* brown = tp-fsdp
* gray = fsdp

loss_train_total:
loss_train_total

throughput_batches_per_sec:
throughput_batches_per_sec

memory_peak_reserved_mem:
memory_peak_reserved_mem

It is okay that we don't see performance improvements here yet -- we'll get those later, in follow up PRs.

### CODE REVIEW

**Comment by eitanturok** [comment]
Currently, the ffn strategy gives different results when we train fsdp vs fsdp-tp. 

See mcli runs:
* mpt-125m-tp-fsdp-IHKG5s
* mpt-125m-fsdp-8OgGjt

and here are their losses which are visibly different.

image


Currently investigating...

**Comment by snarayan21** [comment]
Will review once loss discrepancy has been addressed. good to see it's at least mechanically working though

**Comment by eitanturok** [comment]
brown = tp-fsdp
gray = fsdp
image

The model weights are initialized with different seeds, causing the loss to not match up at the beginning. But we reach the same loss because we end up making the same gradient updates causes the models see the same data.

**Comment by eitanturok** [comment]
* Ignore the mpt-125.yaml -- I'm currently using that for testing
* We test that trainer works with tp in [composer](https://github.com/mosaicml/composer/pull/3620) so I didn't add that test here
* Any other tests we should add here?

File: llmfoundry/command_utils/train.py [hunk 1]
```diff
@@ -502,6 +519,18 @@ def train(cfg: DictConfig) -> Trainer:
...
     _log_num_params(model, logged_cfg)
 
+    # TP config
+    if tp_config is not None:
+        if 'layer_plan' not in tp_config:
+            tp_config['layer_plan'] = {}
+        if 'strategy' in tp_config:
+            strategy = tp_config.pop('strategy')
+            strategy_layer_plan = build_tp_strategy(strategy, model)
+            tp_config['layer_plan'] |= strategy_layer_plan
```
Comment (mvpatel2000) [inline llmfoundry/command_utils/train.py:529]: why is this a |=?

File: llmfoundry/models/utils/tp_strategy.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+def ffn_tp_strategy(model: ComposerModel) -> dict[str, ParallelStyle]:
+    TP_LAYERS = {'up_proj', 'down_proj'}
+
+    # validate that all TP_LAYERS are in model
+    tp_layers_in_model = {
+        layer for layer in TP_LAYERS for name, _ in model.named_modules()
+        if layer in name
+    }
+    assert tp_layers_in_model == TP_LAYERS, f'The FFN tensor parallelism strategy requires `model` to have layers {TP_LAYERS}. But `model` is missing layers {TP_LAYERS - tp_layers_in_model}.'
```
Comment (mvpatel2000) [inline llmfoundry/models/utils/tp_strategy.py:22]: better to do if and then RuntimeError

File: llmfoundry/models/utils/tp_strategy.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    PrepareModuleInput,
+    RowwiseParallel,
+)
+from torch.distributed.tensor.parallel.style import ParallelStyle
+
+
+def ffn_tp_strategy(model: ComposerModel) -> dict[str, ParallelStyle]:
+    TP_LAYERS = {'up_proj', 'down_proj'}
+
+    # validate that all TP_LAYERS are in model
```
Comment (mvpatel2000) [inline llmfoundry/models/utils/tp_strategy.py:17]:
```suggestion
    # Validate that all TP_LAYERS are in model
```

File: llmfoundry/models/utils/tp_strategy.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    tp_layers_in_model = {
+        layer for layer in TP_LAYERS for name, _ in model.named_modules()
+        if layer in name
+    }
+    assert tp_layers_in_model == TP_LAYERS, f'The FFN tensor parallelism strategy requires `model` to have layers {TP_LAYERS}. But `model` is missing layers {TP_LAYERS - tp_layers_in_model}.'
+
+    # generate layer plan
```
Comment (mvpatel2000) [inline llmfoundry/models/utils/tp_strategy.py:24]:
```suggestion
    # Generate layer plan
```

File: llmfoundry/models/utils/tp_strategy.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        elif name.split('.')[-2:] == ['ffn', 'down_proj']:
+            layer_plan[name] = RowwiseParallel(
+                input_layouts=Shard(-1),
+                output_layouts=Shard(0),
+            )
+        elif name.split('.')[-1] == 'ffn':
+            layer_plan[name] = PrepareModuleInput(
+                input_layouts=Shard(0),
+                desired_input_layouts=Replicate(),
+                use_local_output=True,
```
Comment (mvpatel2000) [inline llmfoundry/models/utils/tp_strategy.py:41]: Can we put comments describing what the different shard options here are and what they do? generally describing strategy?

**Review by mvpatel2000 — COMMENTED** [review]
LGTM, a few minor nits.

File: llmfoundry/command_utils/train.py (hunk 1 above)
Comment (eitanturok) [reply llmfoundry/command_utils/train.py:529]: Removed.

Originally we thought the user was going to provide a layer plan and we'd want to combine that with the layer plan of the strategy but then we decided to not support this.

File: llmfoundry/models/utils/tp_strategy.py (hunk 2 above)
Comment (eitanturok) [reply llmfoundry/models/utils/tp_strategy.py:22]: added

**Comment by snarayan21** [comment]
@eitanturok Regarding:
> The model weights are initialized with different seeds
would it be possible to init with the same seed? Granted, even then, loss curves may not match up because of how parameters are initialized on different devices and not sure how TP plays with this...might have to init on cpu or something. Anyways, I'd just check but no need to spend a crazy amount of time trying to get the loss curves to exactly match.

File: llmfoundry/models/utils/tp_strategy.py (hunk 5 above)
Comment (eitanturok) [reply llmfoundry/models/utils/tp_strategy.py]: added comments. can you review them to make sure my explanations are correct?

**Comment by eitanturok** [comment]
> @eitanturok Regarding:
> 
> > The model weights are initialized with different seeds
> 
> would it be possible to init with the same seed? Granted, even then, loss curves may not match up because of how parameters are initialized on different devices and not sure how TP plays with this...might have to init on cpu or something. Anyways, I'd just check but no need to spend a crazy amount of time trying to get the loss curves to exactly match.

We use the same seed in trainer. But behind the scenes, we do seed = seed + rank so that every rank has a unique seed. This is what messes things up. What we really want is seed = seed + tp_group_rank, I think. I looked into this a bit but it got quite complex.

**Review by snarayan21 — COMMENTED** [review]
mostly lgtm, bit more testing

File: llmfoundry/command_utils/train.py [hunk 6]
```diff
@@ -502,6 +519,18 @@ def train(cfg: DictConfig) -> Trainer:
 
     _log_num_params(model, logged_cfg)
 
+    # TP config
+    if tp_config is not None:
+        if 'layer_plan' not in tp_config:
```
Comment (dakinggg) [inline llmfoundry/command_utils/train.py:524]: this is no longer needed, right?

File: llmfoundry/models/__init__.py [hunk 7]
```diff
@@ -24,6 +25,7 @@
 models.register('fmapi_causal_lm', func=FMAPICasualLMEvalWrapper)
 models.register('openai_chat', func=OpenAIChatAPIEvalWrapper)
 models.register('fmapi_chat', func=FMAPIChatAPIEvalWrapper)
+tp_strategy.register('ffn', func=ffn_tp_strategy)
```
Comment (dakinggg) [inline llmfoundry/models/__init__.py:28]: I'd prefer not to stick the tp_strategies in utils. lets make a `llmfoundry/tp` folder, and `tp_strategies.py` in there?

File: llmfoundry/registry.py [hunk 8]
```diff
@@ -389,6 +390,26 @@
     description=_save_planners_description,
 )
 
+_tp_strategy_description = (
```
Comment (dakinggg) [inline llmfoundry/registry.py:393]: the other registries are plural, so lets make this `tp_strategies`?

**Comment by eitanturok** [comment]
Thanks for all the comments, I'm gonna start addressing them!

File: llmfoundry/command_utils/train.py [hunk 9]
```diff
@@ -329,16 +335,27 @@ def train(cfg: DictConfig) -> Trainer:
...
+    if dist.get_world_size(
+    ) == 1 and (fsdp_config is not None or tp_config is not None):
+        parallelism = ''
+        if fsdp_config is not None:
+            parallelism += 'FSDP'
+        if tp_config is not None:
+            parallelism += '+TP' if fsdp_config is not None else 'TP'
         warnings.warn(
-            'FSDP is not applicable for single-GPU training. Reverting to DDP.',
+            f'{parallelism} is not applicable for single-GPU training. Reverting to DDP.',
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply llmfoundry/command_utils/train.py:350]: added

File: llmfoundry/utils/config_utils.py [hunk 10]
```diff
@@ -533,9 +538,14 @@ def process_init_device(model_cfg: dict[str, Any], fsdp_config: Optional[dict]):
             # Set defaults for mixed initialization
             fsdp_config.setdefault('load_monolith_rank0_only', True)
 
+    if tp_config is not None and 'ffn_config' in model_cfg and model_cfg[
+        'ffn_config'].get('ffn_type', None) in ffns_with_megablocks:
+        raise ValueError('Cannot use TP with MoEs.')
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply llmfoundry/utils/config_utils.py]: added test

File: scripts/train/yamls/pretrain/tp-mpt-125m.yaml [hunk 11]
```diff
@@ -0,0 +1,133 @@
+variables:
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply scripts/train/yamls/pretrain/tp-mpt-125m.yaml:1]: I added screenshots to the top comment of the PR. TLDR, no performance improvements on the small 125m-param example.

File: llmfoundry/utils/builders.py [hunk 12]
```diff
@@ -701,3 +703,15 @@ def _validate_cfg(icl_cfg: dict[str, Any]):
                 )
 
     return evaluators, logger_keys
+
+
+def build_tp_strategy(
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply llmfoundry/utils/builders.py:710]: added

File: llmfoundry/command_utils/train.py (hunk 6 above)
Comment (eitanturok) [reply llmfoundry/command_utils/train.py]: yep, I simplified it

File: llmfoundry/registry.py (hunk 8 above)
Comment (eitanturok) [reply llmfoundry/registry.py:393]: done!

File: llmfoundry/models/__init__.py (hunk 7 above)
Comment (eitanturok) [reply llmfoundry/models/__init__.py]: done

**Comment by eitanturok** [comment]
Should we include a `tp-mpt-125m.yaml` in the repo?

**Comment by dakinggg** [comment]
@eitanturok does checkpointing work now?

**Review by mvpatel2000 — APPROVED** [review]
LGTM!

**Comment by mvpatel2000** [comment]
> @eitanturok does checkpointing work now?

No and it won't with FSDPv1.

**Comment by dakinggg** [comment]
@mvpatel2000 @eitanturok ok lets leave the yaml out then

**Comment by dakinggg** [comment]
Also can we log a warning when using TP that checkpointing is known to not work?

File: llmfoundry/command_utils/train.py [hunk 13]
```diff
@@ -18,7 +19,11 @@
     TraceHandler,
     cyclic_schedule,
 )
-from composer.utils import dist, get_device, reproducibility
+from composer.utils import (
+    dist,
+    get_device,
+    reproducibility,
+)
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply llmfoundry/command_utils/train.py]: I put it back onto one line

**Comment by eitanturok** [comment]
@dakinggg I just added a warning that checkpointing does not work + give a link to the exact pytorch issue.

One of the tests verifies that the trainer works but it takes too long cause it downloads a dataset. So I will fix this and I think we will be good to go.
